File: 디자인패턴/Python 디자인 패턴/Iterator/ArrayIterator.py
from Iterator import Iterator
from Aggregator import Aggregator
from My_Item import My_Item

class Array(Aggregator):
    items = []
    
    def __init__(self, items:list)->None: 
        self.items = items
    
    def getItem(self, idx:int)->My_Item:
        return self.items[idx]
    
    def getCount(self):
        return len(self.items)
    
    def iterator(self):
        return ArrayIterator(self)
    pass

class ArrayIterator(Iterator):
    def __init__(self, array:Array):
        self.array = array
        self.index = -1

    def next(self)->bool:
        self.index+=1
        return self.index < self.array.getCount()

    def current(self)->None:
        return self.array.getItem(self.index)
    pass

File: C#/디자인패턴/Iterator/ArrayIterator.cs
public class ArrayIterator : Iterator {
    private Array array;
    private int index;
    public ArrayIterator(Array array) {
        this.array = array;
        this.index = -1;
    }
    public bool next() {
        index++;
        return index < array.getCount();
    }
    public object current() {
        return array.gItem(index);
    }
}
